File: clients/rs/src/features/invoke_wrap_wasm_v0_1.rs
use std::{error::Error, sync::{Arc}, fs, path::Path};
use polywrap_client::{core::{invoker::Invoker, file_reader::SimpleFileReader}, client::PolywrapClient, wasm::wasm_package::WasmPackage, builder::{PolywrapClientConfig, PolywrapClientConfigBuilder}};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::input::{expect_root_dir};

#[derive(Deserialize, Serialize)]
struct Args {
  first: u8,
  second: u8
}

#[derive(Deserialize)]
struct InputObj {
  directory: String,
  method: String,
  args: Args
}

pub fn run_test_case(input: &Value) -> Result<(), Box<dyn Error>> {
  let input_obj: InputObj = serde_json::from_value(input.clone())?;
  let wrap_dir = expect_root_dir(
    &input_obj.directory,
    std::env::current_dir()?.join("../../").to_str().unwrap()
  )?;
  let method = input_obj.method;
  let args = input_obj.args;

  let manifest = fs::read(Path::new(&wrap_dir).join("wrap.info"))?;
  let wasm_module = fs::read(Path::new(&wrap_dir).join("wrap.wasm"))?;

  let wrap_package = WasmPackage::from_bytecode(
    wasm_module,
    Arc::new(SimpleFileReader::new()),
    Some(manifest)
  );

  let mut config = PolywrapClientConfig::new();
  config.add_package("embed/foo".try_into().unwrap(), Arc::new(wrap_package));
  
  let client: PolywrapClient = PolywrapClient::new(config.into());

  println!("Invoking {method}");

  let result = client.invoke_raw(
    &"embed/foo".try_into().unwrap(),
    &method,
    Some(&polywrap_client::msgpack::to_vec(&args)?),
    None,
    None
  );

  match result {
    Ok(result) => {
      let result: u8 = polywrap_client::msgpack::from_slice(&result)?;
      println!("Received: {result:?}");
      println!("Success!");
    },
    Err(e) => {
      println!("{}", e)
    },
  }

  Ok(())
}

File: clients/swift/Sources/Readiness/Features/InvokeWrapWasmV0_1.swift
import Foundation
import PolywrapClient

struct InvokeWrapWasmV0_1Test: Feature {
   func runTestCase(input: Any) throws -> Void {
        guard let inputObj = input as? [AnyHashable: Any] else {
           throw InputError.expectedObject
        }

        let wrapDir = try Input.expectRootDir(inputObj["directory"]!, rootDir: "./")
        let method = try Input.expectString(inputObj["method"]!)

        guard let argsDict = inputObj["args"] as? [AnyHashable: Any],
            let first = argsDict["first"] as? Int,
            let second = argsDict["second"] as? Int else {
            throw InputError.expectedObject
        }

        let args = u8MethodArgs(first: first, second: second)
        let reader = ResourceReader(bundle: Bundle.module)
        let module = try reader.readFile(wrapDir)

        let package = WasmPackage(reader: nil, module: module)
        let uri = try! Uri("embed/foo")

        let builder = BuilderConfig().addPackage(uri, package)

        let client = builder.build()

        print("Invoking \(method)")
        let result: Int? = try? client.invoke(uri: uri, method: method, args: args)
        guard let r = result else {
            return
        }

        print("Received: \(r)")
        print("Success!")
   }
}
